README.md
# `flavored_env`

[![style: very good analysis][very_good_analysis_badge]][very_good_analysis_link]
[![License: MIT][license_badge]][license_link]
[![coverage][coverage_badge]][ci_link]
[![pub package][pub_badge]][pub_link]
[![ci][ci_badge]][ci_link]


Generate and synchronize dotenv and dart configuration classes for your firebase application.

Flavored_env is useful if you need to maintain a common set of constants between your cloud functions and your flutter application. It generates dotenv files for your cloud function deployments as well as corresponding dart class for your application. It supports flavors and the ability to create copies in multiple locations. 

The tools was developed for a personal project to help keep constants in sync during development. Hopefully it can be helpful to others. Do let me know if you are interested in additional features. 

# Features 

* Manage all application and cloud constants in single file
* Create dotenv and Dart classes from configuration
* Supports multiple flavors, copies and locations

## Installation

This package is intended to support development of Dart projects with
[`package:build`][]. In general, put it under [dev_dependencies][], in your
[`pubspec.yaml`][pubspec].

```yaml
dev_dependencies:
  build_runner:
```

### Built-in Commands

The `flavord_env` package exposes a binary by the same name, which can be
invoked using `dart run flavored_env <command>`.

The available commands are

- `create-yaml`: create a sample flavored_env.yaml file.
- `generate`: generate the configuration files based on yaml file
  edits and does rebuilds as necessary.

### flavored_env.yaml 

Create a template ```flavored_env.yaml``` in your project root by running:

```pub run flavored_env --create-yaml```

The default sample will look like this

```yaml
# flavored_env example

output_formats:
  dotenv:
    # Dotenv output options
    build_to:
      - functions
  dart:
    # Dart output options
    build_to:
      - "functions/lib/config"
    class_name: "FlavoredEnv"

  typescript:
    # Typescript output options
    build_to:
      - "functions/src"

flavors:
  default:
    keys:
      COLLECTION_USER_PROFILES:
        value: "users"
        info: collection of user profiles

      COLLECTION_AVATARS:
        value: "avatars"
        info: collection of user avatars

      COLLECTION_USER_SESSIONS:
        value: "sessions"
        info: collection of sessions for user

      DEBUG_DELAY:
        value: 0
        info: A delay in milliseconds used for testing


  staging:
    keys:
      DEBUG_DELAY:
        value: 1000
        info: A delay in milliseconds used for testing

  development:
    keys:
      DEBUG_DELAY:
        value: 4000
        info: A delay in milliseconds used for testing

  production:
    keys:

```

After config is specified in YAML file run following command

```pub run flavored_env --generate```


## Yaml Specification

`dotenv_to` to indicate the list of relative paths where to create the dotenv files.

`class_to` to indicate the list of relative paths where to create the dart files. 

`class_name` to specivy the name of the class to use, default to ```EnvConfig```

`flavors` to specify each flavor or development environment

`default` environment (optional) is used for the base configuration. Its content will be used to generate _.env.default_. When generating the dart files for other flavors, the keys from the default environment that are not overriden are merged to make sure all constants are available to your flutter app. 

`keys` to specify the list of constants for the specific environment. Each key is specified by its name and can take two attributes; `value` that represents the constant value to use and `info` that provides one line descriptions that will be used as comments in the generated files. 

Upon generation, the key names will be converted to camel case for the dart files and upper case separated by underscore for the dotenv to match normal conventions. 

Any `environment` listed in the file will translate into _.env._`environment` and env_config_`environment`.dart in each of the specified locations. 


### Limitations

* Destination paths are relative only. No absolute paths
* Destination paths must exist and are not created
* Opiniated format for keys in dart and dotenv for now


[ci_badge]: https://github.com/berbsd/flavored_env/actions/workflows/flavored_env.yaml/badge.svg?branch=main
[ci_link]: https://github.com/berbsd/flavored_env/actions
[coverage_badge]: https://raw.githubusercontent.com/berbsd/flavored_env/main/coverage_badge.svg
[pub_badge]: https://img.shields.io/pub/v/flavored_env.svg
[pub_link]: https://pub.dartlang.org/packages/flavored_env
[license_badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license_link]: https://opensource.org/licenses/MIT






[license_badge]: https://img.shields.io/badge/license-MIT-blue.svg
[license_link]: https://opensource.org/licenses/MIT
[very_good_analysis_badge]: https://img.shields.io/badge/style-very_good_analysis-B22C89.svg
[very_good_analysis_link]: https://pub.dev/packages/very_good_analysis
[mason]: https://pub.dev/packages/mason_cli




CHANGELOG.md
## 0.6.0-dev.2
- Downgrade path package to be compatible with flutter_test
## 0.6.0-dev.1
- Rewrite using mason brick console_program
## 0.5.0
- Add typescript file to access all the dotenv files. 
- Code refactoring
## 0.4.0
- Use functions instead of static for Dart

## 0.3.0
- Add version check and warn user if outdated

## 0.2.0
- Minor fixes
- Updated documentation and tests 

## 0.1.1
- Updated description for publication
## 0.1.0
- No functional changes.
- Updated package versions for compatibility

## 0.0.1
- Initial version.


lib/src/config/config.dart
/// default dart class name if none provided in yaml file
const String dartClassName = 'EnvConfig';

/// default output path for dart class if none provided in yaml file
List<String> dartBuildTo = ['lib'];

/// default output path for dart class if none provided in yaml file
List<String> tsBuildTo = ['functions/src'];

/// name of the key field in yaml file
const String keysField = 'keys';

/// name of the value field in yaml file
const String valueField = 'value';

/// name of the build_to field in yaml file
const String buildToField = 'build_to';

/// name of the info field in yaml file
const String infoField = 'info';

/// name of the output_format field in yaml file
const String outputFormatsField = 'output_formats';

/// name of the flavors field in yaml file
const String flavorsField = 'flavors';

/// defauilt filename for typesctript constants
const String flavoredEnvFile = 'flavored_env';

/// defauilt typesctript file extension
const String tsExtension = '.g.ts';

/// defauilt dart file extension
const String dartExtension = '.g.dart';

/// default config file header
String configFileHeader({
  required String comment,
  required String flavorName,
}) =>
    '''
$comment AUTO-GENERATED - DO NOT MODIFY THIS FILE
$comment ----------------------------------------
$comment FLAVOR: $flavorName

''';

/// default yaml filename
const String yamlTemplateName = 'flavored_env.yaml';

/// template yaml content to kickstart customization
const String yamlTemplateContent = '''
output_formats:
  dotenv:
    # Dotenv output options
    build_to:
      - functions
  dart:
    # Dart output options
    build_to:
      - "lib/config"
    class_name: "EnvConfig"

  typescript:
    # Typescript output options
    build_to:
      - "functions/src"

flavors:
  default:
    keys:
      DEFAULT_TARGET_HOURS:
        value: 16
        info: Default target hours

      COLLECTION_USER_PROFILES:
        value: "users"
        info: collection of user profiles

      COLLECTION_AVATARS:
        value: "avatars"
        info: collection of user avatars

      COLLECTION_APP_USERS:
        value: "app-users"
        info: collection for app-wide information

      DOCUMENT_APP_USERS_TOTAL:
        value: "users-total"
        info: document holding total users

      COLLECTION_TRACKERS:
        value: "trackers"
        info: collection of active trackers

      DOCUMENT_APP_USERS_FASTING:
        value: "users-fasting"
        info: document holding total fasting users

      COLLECTION_USER_SESSIONS:
        value: "sessions"
        info: collection of sessions for user

      DEBUG_DELAY:
        value: 0
        info: A delay in milliseconds used for testing

      SLACK_WEBHOOK_BASE_URL:
        value: "https://hooks.slack.com/services/T037QP90QN9"
        info: Slack webhook base URL

  staging:
    keys:
      SLACK_SIGNUPS_WEBHOOK:
        value: "B038C5B2A15/3ysdlkf5SrJOF2pqoGvV078gHB"
        info: Slack webhook for user sign-up notification

      DEBUG_DELAY:
        value: 1000
        info: A delay in milliseconds used for testing

  development:
    keys:
      SLACK_SIGNUPS_WEBHOOK:
        value: "GEW7QTHPW77/SQqmocFslk32ORQXqR7qyqI"
        info: Slack webhook for user sign-up notification

      DEBUG_DELAY:
        value: 4000
        info: A delay in milliseconds used for testing

  production:
    keys:
''';
